### PR TITLE
Retry GraphQL queries with backoff

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- retry GraphQL requests up to five times using exponential backoff with jitter
- add rand dependency for jitter and document networking retries
- test retry behaviour on missing data responses
- verify review thread pagination retries preserve order and skip duplicates
- document retry policy and back-off jitter in design guide

## Testing
- `make fmt`
- `make lint`
- `make test`
- `make markdownlint`
- `make nixie` *(fails: Error running command bun x --bun @mermaid-js/mermaid-cli ... FileNotFound)*

------
https://chatgpt.com/codex/tasks/task_e_689dc6305d3c83228ea414a60b6abe4d

## Summary by Sourcery

Implement retry logic for GraphQL client with exponential backoff and jitter, update tests and documentation accordingly

New Features:
- Retry GraphQL run_query up to five times on network errors or missing data responses using exponential backoff with random jitter
- Preserve pagination order and avoid duplicates when retrying review thread pagination

Enhancements:
- Add should_retry filter with constants for retry attempts and base delay and integrate rand for jitter generation

Build:
- Add rand 0.8 dependency for jitter

Documentation:
- Document the retry policy and backoff jitter in the design guide

Tests:
- Add tests for retry behaviour on missing data responses
- Add tests to verify pagination retries preserve order and skip duplicates

### CODE REVIEW

File: src/api/mod.rs [hunk 1]
```diff
@@ -125,78 +129,111 @@ impl GraphQLClient {
...
+                    Err(e) => {
+                        let snippet = snippet(
+                            &serde_json::to_string_pretty(&value).unwrap_or_default(),
+                            VALUE_SNIPPET_LEN,
+                        );
+                        let path = e.path().to_string();
+                        let inner = e.into_inner();
+                        Err(VkError::BadResponseSerde(
+                            format!("{inner} at {path} | snippet: {snippet}").boxed(),
+                        ))
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/api/mod.rs:218]: @coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix.

File: src/api/mod.rs [hunk 2]
```diff
@@ -256,9 +293,78 @@ where
 
 #[cfg(test)]
 mod tests {
+    //! Tests for API utilities.
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/api/mod.rs]: Has this now been resolved?

File: src/api/mod.rs [hunk 3]
```diff
@@ -125,78 +129,111 @@ impl GraphQLClient {
...
+        for attempt in 0..RETRY_ATTEMPTS {
+            let result = async {
+                let response = self
+                    .client
+                    .post(&self.endpoint)
+                    .headers(self.headers.clone())
+                    .json(&payload)
+                    .send()
+                    .await
+                    .map_err(|e| VkError::RequestContext {
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/api/mod.rs]: Has this now been resolved?

File: src/api/mod.rs [hunk 4]
```diff
@@ -125,78 +129,111 @@ impl GraphQLClient {
...
     ///
     /// # Errors
     ///
     /// Returns a [`VkError`] if the request fails or the response cannot be
     /// deserialised.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the retry base delay exceeds `u64` milliseconds.
     pub async fn run_query<V, T>(&self, query: &str, variables: V) -> Result<T, VkError>
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/api/mod.rs:261]: Has this now been resolved?
